### ISSUE TITLE
loss function

### ISSUE THREAD

**Issue by Baekhyunn** [fengyanzi/FAT#2]
您将会释放您的损失函数或者训练代码吗？期待您的回复，感恩！

**Comment by fengyanzi** [comment]
请问您是中文读者 or are you English Reader？我计划上传代码，但是需要一点时间最近我很忙，刚刚中稿了一篇CVPR，本科生毕业设计还没写，这些事情需要得到优先解决。至于训练代码，我的代码是基于https://github.com/fengyanzi/MRF-Net微调而来，大量变量的命名可能存在误导性，修改需要大量时间，如果你需要可以对MRF-Net的训练函数进行微调。在此，我强烈推荐您使用BasicSR库，其中包含大量SR研究需要的训练方法与损失函数等工具，本项目并未采用BasicSR是一个遗憾，仅管BasicSR上手难度确实大，兼容性存在一定问题，但如果你能熟练掌握，者能节约您的大量时间（介于一些SOTA算法均由BasicSR搭建）

训练代码（包含大量错误）：
`
import os
import torch.nn as nn
import torch
from torchvision import transforms
from torch.utils.data import DataLoader, Dataset
from PIL import Image
import argparse
import random
import numpy as np
from tqdm import tqdm
from models.FAT import FAT 
from datetime import datetime
from torchvision import models
import torch.nn.functional as f

class CloudRemovalDataset(Dataset):
    def __init__(self, root_dir):
        self.root_dir = root_dir
        self.clear_dir = os.path.join(root_dir, 'clear')
        self.cloud_dir = os.path.join(root_dir, 'down')
        self.image_filenames = [img for img in os.listdir(self.cloud_dir) if
                                img.endswith('.png') or img.endswith('.jpg')]
        self.transform1 = transforms.Compose([
            transforms.Resize(120),
            transforms.ToTensor(),
            transforms.Normalize(mean=[0.5,0.5,0.5], std=[0.5,0.5,0.5])])
        self.transform2 = transforms.Compose([
            transforms.Resize(480),
            transforms.ToTensor(),
            transforms.Normalize(mean=[0.5,0.5,0.5], std=[0.5,0.5,0.5])])

    def __len__(self):
        return len(self.image_filenames)

    def __getitem__(self, idx):
        cloud_img_name = os.path.join(self.cloud_dir, self.image_filenames[idx])
        clear_img_name = os.path.join(self.clear_dir, self.image_filenames[idx])
        cloud_img = Image.open(cloud_img_name) #.convert('L')
        clear_img = Image.open(clear_img_name) #.convert('L')
        clear_img = self.transform2(clear_img)
        cloud_img = self.transform1(cloud_img)
        sample = {'cloud_img': cloud_img, 'clear_img': clear_img}
        return sample
def train(model, train_dataloader, optimizer, criterion,criterion1, device, save_dir, save_cycle, epochs, resume=None,
          crop_size=120, texturesize=20):
    if resume is not None:
        if not os.path.exists(resume):
            raise FileNotFoundError(f'Checkpoint {resume} not found.')
        else:
            print(f'Resume from {resume}')
            model.load_state_dict(torch.load(args.resume))
    else:
        print('Train from scratch...')
    if not os.path.exists(save_dir):
        os.makedirs(save_dir)
    best_loss = float("inf")

    content_loss = PerceptualLoss(torch.nn.MSELoss())
    content_loss.contentFunc.to(device)
    gradient_loss = Gradient_Net().to(device)
    L1_Charbonnier = L1_Charbonnier_loss()
    CustomLosses = CustomLoss(20).to(device)
    #CustomLosses_imp = impCustomLoss(10,40).to(device)
    CustomLosses_imp = CustomPerceptualLoss(10, 40).to(device)
    #fourierloss=FourierL1Loss().to(device)
    VarFeatureLoss = VariancePerceptualLoss().to(device)
    VarimpFeatureLoss = VarianceimpFeatureLoss().to(device)







    for epoch in range(epochs):
        losses = {'loss': [], 'loss1': [], 'loss2': [],'loss3': [],'loss4': [],'loss5': [],'loss6': [],'loss7': []}
        for batch in tqdm(train_dataloader):
            cloud_imgs = batch['cloud_img'].to(device)
            clear_imgs = batch['clear_img'].to(device)
            _, _, height, width = cloud_imgs.shape
            #if crop_size >= width or crop_size >= height:
            #    raise ValueError('crop_size must be less than img_size')
            #max_coordinate = min(width, height) - crop_size
            #x = random.randint(0, max_coordinate)
            #y = random.randint(0, max_coordinate)
            #cropped_cloud_imgs = cloud_imgs[:, :, y:y + crop_size, x:x + crop_size]
            #x1 = random.randint(texturesize / 2, min(width, height) - texturesize / 2)
            #y1 = random.randint(texturesize / 2, min(width, height) - texturesize / 2)
            #texture_cloud_imgs = cloud_imgs[:, :, int(y1 - texturesize / 2):int(y1 + texturesize / 2),
            #                     int(x1 - texturesize / 2):int(x1 + texturesize / 2)]

            res = model(cloud_imgs)
            #cropped_res = model(cropped_cloud_imgs)
            #texture_res = model(texture_cloud_imgs)
            loss1 = criterion(res, clear_imgs)
            loss3 = content_loss.get_loss(res, clear_imgs)
            #loss4=gradient_loss.get_loss(res, clear_imgs)


            loss4=L1_Charbonnier.forward(res, clear_imgs)


            loss5=CustomLosses(res, clear_imgs)
            loss6 = CustomLosses_imp(res, clear_imgs)
            #loss7 = fourierloss.forward(res, clear_imgs)
            loss7 = VarimpFeatureLoss.forward(res, clear_imgs)

            #loss2=criterion1(res, clear_imgs)
            #loss2 = criterion(res[:, :, y:y + crop_size, x:x + crop_size], cropped_res)
            loss2 = criterion1(res, clear_imgs)
            loss = loss1 +loss3*0.001#+0.1*loss7#+0.01*loss6#+0.1*loss6#+1 *loss2+loss1++loss4
            loss.backward()
            optimizer.step()
            optimizer.zero_grad()
            losses['loss'].append(loss.item())
            losses['loss1'].append(loss1.item())
            losses['loss2'].append(loss2.item())
            losses['loss3'].append(loss3.item())
            losses['loss4'].append(loss4.item())
            losses['loss5'].append(loss5.item())
            losses['loss6'].append(loss6.item())
            losses['loss7'].append(loss7.item())
        current_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

        print('{} | Epoch:{}/{} | Loss:{:.5f} Loss1:{:.5f} Loss2:{:.5f}Loss3:{:.5f}Loss4:{:.5f}Loss5:{:.5f}Loss6:{:.5f}Loss7:{:.5f}'
              .format(current_time, epoch + 1, epochs, np.mean(losses['loss']), np.mean(losses['loss1']), np.mean(losses['loss2']),np.mean(losses['loss3']),np.mean(losses['loss4']),np.mean(losses['loss5']),np.mean(losses['loss6']),np.mean(losses['loss7'])
                      ))
        content = '{} | Epoch:{}/{} | Loss:{:.5f} Loss1:{:.5f}Loss2:{:.5f}Loss3:{:.5f}Loss4:{:.5f}Loss5:{:.5f}Loss6:{:.5f}Loss7:{:.5f}'.format(current_time,
                                                                                                 epoch + 1, epochs,
                                                                                                   np.mean(
                                                                                                     losses['loss']),
                                                                                                 np.mean(
                                                                                                     losses['loss1']),
                                                                                   np.mean(
                                                                                       losses['loss2']),
                                                                                   np.mean(
                                                                                       losses['loss3']),
                                                                                                           np.mean(losses['loss4']),np.mean(losses['loss5']),np.mean(losses['loss6']),np.mean(losses['loss7'])

                                                                                   )
        file_path_log = os.path.join(args.save_dir, 'output.txt')
        with open(file_path_log, 'a') as file:
            # 写入内容并换行
            file.write(content + '\n')

        if np.mean(losses['loss']) <= best_loss:
            torch.save(model.state_dict(), os.path.join(save_dir, 'best.pth'))
            best_loss = np.mean(losses['loss'])
            #print('Saving best model...')
        if (epoch + 1) % save_cycle == 0:
            torch.save(model.state_dict(), os.path.join(save_dir, '{}.pth'.format(epoch + 1)))
    print('\nTrain Complete.\n')
    #torch.save(model.state_dict(), os.path.join(save_dir, 'last.pth'))


def main(args):
    dataset = CloudRemovalDataset(root_dir=args.data_dir)
    #dataset = RealdownsampleDataset(root_dir=args.data_dir)
    train_dataloader = DataLoader(dataset, batch_size=args.batch_size, shuffle=True, num_workers=8)
    device = torch.device("cuda" if torch.cuda.is_available() and not args.no_cuda else "cpu")
    print(f'Using device: {device}')
    # model = IRDecloud().to(device)
    #model = MRDA_net().to(device)
    model=FAT().to(device)
    # model = FFA(gps=3,blocks=19).to(device)
    criterion = nn.L1Loss().to(device)
    criterion1=nn.MSELoss().to(device)



    optimizer = torch.optim.Adam(params=filter(lambda x: x.requires_grad, model.parameters()), lr=args.lr,
                                 betas=(0.9, 0.999),
                                 eps=1e-08)
    optimizer.zero_grad()
    subtitle = """  
        ______               __     ___  
       / ____/ ____ _ _____ / /_ __/ /_____  _,___ ,_______ ____  _,___
      / /__   / __ `// ___// __//_  _// __ \|  __// _  _  // __ \|  __/
     / /     / /_/ /(__  )/ /_  _/ / / /_/ // /  / // // // /___// /  
    /_/      \__,_//____/ \__/ (__/  \____//_/  /_//_//_/ \___/ /_/  
         ______                   __   __                 __      __  
        / ____/____   ____   ____/ /  / /   __  __ _____ / /__   / /  
       / / __ / __ \ / __ \ / __  /  / /   / / / // ___// //_/  / /  
      / /_/ // /_/ // /_/ // /_/ /  / /___/ /_/ // /__ / /<    /_/  
      \____/ \____/ \____/ \____/  /_____/\____/ \___//_/|_|  (_)  
    """

    # 使用print函数打印这个字符串
    print(subtitle)
    train(model, train_dataloader, optimizer, criterion,criterion1, device, args.save_dir, args.save_cycle, args.epochs,
          args.resume)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Train a model.')#D:\yxylearn\IRfenbian\datasets\Sen480\cali D:\yxylearn\test\data\test\all  D:\yxylearn\test\data\test\BIT-SR24   D:\yxylearn\test\data\df2k
    parser.add_argument('--data_dir', type=str, default=r'D:\yxylearn\test\data\test\BIT-SR24', help='Path to dataset')
    parser.add_argument('--save_dir', type=str, default=r'./version/checkpoints6/realv1fewXU', help='Path to save checkpoints')
    parser.add_argument('--save_cycle', type=int, default=5, help='Cycle of saving checkpoint')
    parser.add_argument('--resume',default=r'./version/checkpoints6/realv1few/best.pth', type=str,help='Path to checkpoint file to resume training')
    parser.add_argument('--lr', type=float, default=0.0001, help='Learning rate')
    parser.add_argument('--batch_size', type=int, default=2, help='Batch size of each data batch')
    parser.add_argument('--no-cuda', action='store_true', default=False, help='Disable CUDA')
    parser.add_argument('--epochs', type=int, default=10, help='Epochs')
    args = parser.parse_args()
    main(args)
`

**Comment by Baekhyunn** [comment]
我是个中文读者，对您的损失函数的具体设计很感兴趣。我的超分项目为基于BasicSR框架编写，如果您有需要帮忙整理也可以联系我。感谢您的回答。

**Comment by Baekhyunn** [comment]
hh感谢

**Comment by zhuguangchun1** [comment]
真厉害，一个即将入学的硕士已发顶会！

**Comment by fengyanzi** [comment]
谢谢
